vcl/main.vcl

```diff
@@ -161,8 +161,115 @@ sub vcl_recv {
 	}
 
 	if (req.url ~ "^/v2/recordRumData" && req.http.Normalized-User-Agent) {
-		set req.http.Log = regsub(req.url, "^.*?\?(.*)$", "\1") "&ip=" client.ip "&refer_domain=" regsub(req.http.Referer, "^(https?\:\/\/)?(www\.)?(.+?)(\:\d+)?([\/\?].*)?$", "\3") "&country=" geoip.country_code "&data_center=" if(req.http.Cookie:FastlyDC, req.http.Cookie:FastlyDC, server.datacenter);
-		error 204 "No Content";
+		declare local var.rumRequestID STRING;
+		declare local var.rumLogString STRING;
+		declare local var.hashSalt STRING;
+		declare local var.safeIP INTEGER;
+		declare local var.safeRef STRING;
+
+		set var.hashSalt = "salt";
+		set var.rumRequestID = now.sec "-" randomstr(10, "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789");
+		set var.safeIP = addr.extract_bits(client.ip, 16, 32);
+		set var.safeRef = regsub(req.http.Referer, "^(https?\:\/\/)?(www\.)?(.+?)(\:\d+)?([\/\?].*)?$", "\3");
+		set var.rumLogString = "{"
+
+			# Strings
+			{""request_id":""} cstr_escape(var.rumRequestID) {"","}
+			{""request_time":""} strftime({"%Y-%m-%dT%H:%M:%SZ"}, time.start) {"","}
+			{""country":""} cstr_escape(geoip.country_code) {"","}
+			{""data_center":""} cstr_escape(if(req.http.Cookie:FastlyDC, req.http.Cookie:FastlyDC, server.datacenter)) {"","}
+			{""refer_domain_hash":""} cstr_escape(regsub(digest.hash_sha256(var.hashSalt var.safeRef), "^(.{15}).*?$", "\1")) {"","}
+			{""client_ip_hash":""} cstr_escape(regsub(digest.hash_sha256(var.hashSalt var.safeIP), "^(.{15}).*?$", "\1")) {"","}
+			{""ua_family":""} cstr_escape(regsub(urldecode(req.http.Normalized-User-Agent), "^(\w+)\/.*$", "\1")) {"","}
+
+			# Integers
+			# CAREFUL: These variables are concatenated into a raw JSON string.
+			# Ensure they are valid numbers
+			{""perf_dns":"} regsub(req.url.qs, "^(.*\&)?dns=(\d+).*?$", "\2") ","
```

Comment:
Should we use `std.atoi` here?

Reply:
No reason to cast it to an int when we actually need a string anyway.  But there's a hole here: if the regex doesn't match you end up dumping the entire querystring into the output.  I'm refactoring it.